File: apps/mobile/lib/core/split_math.dart
/// Dart port of `packages/shared/src/splitMath.ts`.
///
/// The API recomputes stored allocations with the TypeScript original, so this
/// port must match it exactly (integer-cents math, per-item even split with
/// leftover cents rotated by item index, extras via largest-remainder or even).
library;

class ItemizedLineItem {
  final double total;
  final List<String> assignedMemberIds;

  const ItemizedLineItem({
    required this.total,
    required this.assignedMemberIds,
  });
}

class ItemizedAllocationDetail {
  final String memberId;
  final double itemsAmount;
  final double extrasAmount;
  final double amount;

  const ItemizedAllocationDetail({
    required this.memberId,
    required this.itemsAmount,
    required this.extrasAmount,
    required this.amount,
  });
}

class ItemizedAllocationResult {
  final List<ItemizedAllocationDetail> allocations;
  final double itemsSubtotal;
  final double extrasTotal;
  final double grandTotal;

  const ItemizedAllocationResult({
    required this.allocations,
    required this.itemsSubtotal,
    required this.extrasTotal,
    required this.grandTotal,
  });
}

double roundCents(double value) => (value * 100).roundToDouble() / 100;

int _toCents(double value) => (value * 100).round();

/// Converts per-line-item member assignments into cent-accurate per-person
/// allocations. Each item's cost is split evenly among the members assigned to
/// it; tax + tip are then layered on top either proportionally to each
/// person's item subtotal or evenly across everyone with an assignment.
/// The returned allocation amounts always sum exactly to
/// items subtotal + tax + tip.
ItemizedAllocationResult buildItemizedAllocations({
  required List<ItemizedLineItem> lineItems,
  double tax = 0,
  double tip = 0,
  String extrasSplitMode = 'proportional',
}) {
  // Member order is first appearance across items so results are stable.
  final memberOrder = <String>[];
  final itemCentsByMember = <String, int>{};
  void track(String memberId) {
    if (!itemCentsByMember.containsKey(memberId)) {
      itemCentsByMember[memberId] = 0;
      memberOrder.add(memberId);
    }
  }

  var itemsSubtotalCents = 0;
  for (var itemIndex = 0; itemIndex < lineItems.length; itemIndex += 1) {
    final item = lineItems[itemIndex];
    final assigned = item.assignedMemberIds;
    if (assigned.isEmpty) continue;
    final cents = _toCents(item.total);
    itemsSubtotalCents += cents;

    final base = cents ~/ assigned.length;
    var remainder = cents - base * assigned.length;
    for (final memberId in assigned) {
      track(memberId);
    }
    // Rotate who absorbs leftover cents by item index so no single member
    // is systematically overcharged across a long receipt.
    for (var i = 0; i < assigned.length; i += 1) {
      final memberId = assigned[(i + itemIndex) % assigned.length];
      var share = base;
      if (remainder > 0) {
        share += 1;
        remainder -= 1;
      }
      itemCentsByMember[memberId] = (itemCentsByMember[memberId] ?? 0) + share;
    }
  }

  final extrasCents = _toCents(tax) + _toCents(tip);
  final extrasByMember = <String, int>{};

  if (memberOrder.isNotEmpty && extrasCents != 0) {
    if (extrasSplitMode == 'even' || itemsSubtotalCents == 0) {
      final base = extrasCents ~/ memberOrder.length;
      var remainder = extrasCents - base * memberOrder.length;
      for (final memberId in memberOrder) {
        var share = base;
        if (remainder > 0) {
          share += 1;
          remainder -= 1;
        }
        extrasByMember[memberId] = share;
      }
    } else {
      // Largest-remainder method: floor each proportional share, then hand
      // leftover cents to the members with the biggest truncated fraction.
      var assignedCents = 0;
      final fractions = <({String memberId, int index, double fraction})>[];
      for (var index = 0; index < memberOrder.length; index += 1) {
        final memberId = memberOrder[index];
        final itemCents = itemCentsByMember[memberId] ?? 0;
        final exact = (extrasCents * itemCents) / itemsSubtotalCents;
        final floored = exact.floor();
        assignedCents += floored;
        extrasByMember[memberId] = floored;
        fractions.add((
          memberId: memberId,
          index: index,
          fraction: exact - floored,
        ));
      }
      var leftover = extrasCents - assignedCents;
      fractions.sort((a, b) {
        final byFraction = b.fraction.compareTo(a.fraction);
        if (byFraction != 0) return byFraction;
        return a.index.compareTo(b.index);
      });
      for (final entry in fractions) {
        if (leftover <= 0) break;
        extrasByMember[entry.memberId] =
            (extrasByMember[entry.memberId] ?? 0) + 1;
        leftover -= 1;
      }
    }
  }

  final allocations = memberOrder.map((memberId) {
    final itemsCents = itemCentsByMember[memberId] ?? 0;
    final extras = extrasByMember[memberId] ?? 0;
    return ItemizedAllocationDetail(
      memberId: memberId,
      itemsAmount: itemsCents / 100,
      extrasAmount: extras / 100,
      amount: (itemsCents + extras) / 100,
    );
  }).toList();

  return ItemizedAllocationResult(
    allocations: allocations,
    itemsSubtotal: itemsSubtotalCents / 100,
    extrasTotal: extrasCents / 100,
    grandTotal:
        (itemsSubtotalCents + (memberOrder.isNotEmpty ? extrasCents : 0)) / 100,
  );
}

File: packages/shared/src/splitMath.ts
import type { ExtrasSplitMode } from "./types.js";

export interface ItemizedLineItem {
  total: number;
  assignedMemberIds: string[];
}

export interface ItemizedAllocationInput {
  lineItems: ItemizedLineItem[];
  tax?: number;
  tip?: number;
  extrasSplitMode?: ExtrasSplitMode;
}

export interface ItemizedAllocationDetail {
  memberId: string;
  itemsAmount: number;
  extrasAmount: number;
  amount: number;
}

export interface ItemizedAllocationResult {
  allocations: ItemizedAllocationDetail[];
  itemsSubtotal: number;
  extrasTotal: number;
  grandTotal: number;
}

export const roundCents = (value: number): number =>
  Math.round((value + Number.EPSILON) * 100) / 100;

const toCents = (value: number): number => Math.round(value * 100);

/**
 * Converts per-line-item member assignments into cent-accurate per-person
 * allocations. Each item's cost is split evenly among the members assigned to
 * it; tax + tip are then layered on top either proportionally to each
 * person's item subtotal or evenly across everyone with an assignment.
 * The returned allocation amounts always sum exactly to
 * items subtotal + tax + tip.
 *
 * Single source of truth: the API derives stored allocations from this and
 * the web form previews with it, so both sides always agree to the cent.
 */
export const buildItemizedAllocations = (
  input: ItemizedAllocationInput
): ItemizedAllocationResult => {
  const { lineItems, tax = 0, tip = 0, extrasSplitMode = "proportional" } = input;

  // Member order is first appearance across items so results are stable.
  const memberOrder: string[] = [];
  const itemCentsByMember = new Map<string, number>();
  const track = (memberId: string) => {
    if (!itemCentsByMember.has(memberId)) {
      itemCentsByMember.set(memberId, 0);
      memberOrder.push(memberId);
    }
  };

  let itemsSubtotalCents = 0;
  lineItems.forEach((item, itemIndex) => {
    const assigned = item.assignedMemberIds;
    if (!assigned.length) return;
    const cents = toCents(item.total);
    itemsSubtotalCents += cents;

    const base = Math.floor(cents / assigned.length);
    let remainder = cents - base * assigned.length;
    assigned.forEach((memberId) => track(memberId));
    // Rotate who absorbs leftover cents by item index so no single member
    // is systematically overcharged across a long receipt.
    for (let i = 0; i < assigned.length; i += 1) {
      const memberId = assigned[(i + itemIndex) % assigned.length];
      let share = base;
      if (remainder > 0) {
        share += 1;
        remainder -= 1;
      }
      itemCentsByMember.set(
        memberId,
        (itemCentsByMember.get(memberId) ?? 0) + share
      );
    }
  });

  const extrasCents = toCents(tax) + toCents(tip);
  const extrasByMember = new Map<string, number>();

  if (memberOrder.length > 0 && extrasCents !== 0) {
    if (extrasSplitMode === "even" || itemsSubtotalCents === 0) {
      const base = Math.floor(extrasCents / memberOrder.length);
      let remainder = extrasCents - base * memberOrder.length;
      memberOrder.forEach((memberId) => {
        let share = base;
        if (remainder > 0) {
          share += 1;
          remainder -= 1;
        }
        extrasByMember.set(memberId, share);
      });
    } else {
      // Largest-remainder method: floor each proportional share, then hand
      // leftover cents to the members with the biggest truncated fraction.
      let assignedCents = 0;
      const fractions = memberOrder.map((memberId, index) => {
        const itemCents = itemCentsByMember.get(memberId) ?? 0;
        const exact = (extrasCents * itemCents) / itemsSubtotalCents;
        const floored = Math.floor(exact);
        assignedCents += floored;
        extrasByMember.set(memberId, floored);
        return { memberId, index, fraction: exact - floored };
      });
      let leftover = extrasCents - assignedCents;
      fractions.sort((a, b) => b.fraction - a.fraction || a.index - b.index);
      for (const entry of fractions) {
        if (leftover <= 0) break;
        extrasByMember.set(
          entry.memberId,
          (extrasByMember.get(entry.memberId) ?? 0) + 1
        );
        leftover -= 1;
      }
    }
  }

  const allocations = memberOrder.map((memberId) => {
    const itemsCents = itemCentsByMember.get(memberId) ?? 0;
    const extras = extrasByMember.get(memberId) ?? 0;
    return {
      memberId,
      itemsAmount: itemsCents / 100,
      extrasAmount: extras / 100,
      amount: (itemsCents + extras) / 100
    };
  });

  return {
    allocations,
    itemsSubtotal: itemsSubtotalCents / 100,
    extrasTotal: extrasCents / 100,
    grandTotal:
      (itemsSubtotalCents + (memberOrder.length ? extrasCents : 0)) / 100
  };
};

/**
 * Splits a printed quantity line ("4 Breakfast Hash  68.00") into per-unit
 * amounts that sum exactly to the line total — the first units absorb any
 * leftover cents.
 */
export const splitTotalIntoUnits = (
  total: number,
  quantity: number
): number[] => {
  if (quantity <= 1) return [total];
  const totalCents = Math.round(total * 100);
  const base = Math.floor(totalCents / quantity);
  let remainder = totalCents - base * quantity;
  return Array.from({ length: quantity }, () => {
    let cents = base;
    if (remainder > 0) {
      cents += 1;
      remainder -= 1;
    }
    return cents / 100;
  });
};
